File: android/app/src/main/java/dev/phylliida/anki/SaveFolderPlugin.java
// Capacitor plugin: let the user pick a folder (SAF document tree) and
// read/write individual files inside it. oss-anki keeps its whole collection
// in `oss-anki.json` there so external tools (Syncthing etc.) can sync it.
//
// The picked tree URI is persisted (takePersistableUriPermission) and stored
// in SharedPreferences, so the choice survives restarts. Writes go to a
// `<name>.tmp` sibling that is then renamed over the target — a crash
// mid-write can never leave a truncated save file, and file watchers only
// ever see complete files.
//
// Contract with web/native-bridge.js:
//   pickFolder()            -> { uri, label }          rejects "CANCELLED"
//   getFolder()             -> { uri, label } | { uri: null }
//   readFile({ name })      -> { data: base64 | null }
//   writeFile({ name, data: base64 }) -> { modified }  (atomic tmp+rename)
//   statFile({ name })      -> { exists, modified, size }
//   exportFile({ name, data: base64, mimeType }) -> { uri }
//                             "save as" picker (ACTION_CREATE_DOCUMENT);
//                             rejects "CANCELLED"
//
// Registered from MainActivity.onCreate via registerPlugin(...).

package dev.phylliida.anki;

import android.content.Intent;
import android.content.SharedPreferences;
import android.net.Uri;
import android.util.Base64;

import androidx.activity.result.ActivityResult;
import androidx.documentfile.provider.DocumentFile;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.ActivityCallback;
import com.getcapacitor.annotation.CapacitorPlugin;

import java.io.ByteArrayOutputStream;
import java.io.InputStream;
import java.io.OutputStream;

@CapacitorPlugin(name = "SaveFolder")
public class SaveFolderPlugin extends Plugin {

    private static final String PREFS = "SaveFolderPrefs";
    private static final String KEY_URI = "treeUri";
    private static final String KEY_LABEL = "treeLabel";

    private SharedPreferences prefs() {
        return getContext().getSharedPreferences(PREFS, 0);
    }

    private DocumentFile tree() {
        String uri = prefs().getString(KEY_URI, null);
        if (uri == null) return null;
        return DocumentFile.fromTreeUri(getContext(), Uri.parse(uri));
    }

    @PluginMethod
    public void pickFolder(PluginCall call) {
        Intent intent = new Intent(Intent.ACTION_OPEN_DOCUMENT_TREE);
        intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION
                | Intent.FLAG_GRANT_WRITE_URI_PERMISSION
                | Intent.FLAG_GRANT_PERSISTABLE_URI_PERMISSION);
        startActivityForResult(call, intent, "pickFolderResult");
    }

    @ActivityCallback
    private void pickFolderResult(PluginCall call, ActivityResult result) {
        if (call == null) return;
        Intent data = result.getData();
        Uri uri = data != null ? data.getData() : null;
        if (uri == null) {
            call.reject("Folder pick cancelled", "CANCELLED");
            return;
        }
        try {
            final int flags = Intent.FLAG_GRANT_READ_URI_PERMISSION
                    | Intent.FLAG_GRANT_WRITE_URI_PERMISSION;
            getContext().getContentResolver().takePersistableUriPermission(uri, flags);
        } catch (SecurityException ignored) {
            // Some providers don't offer persistable grants; the URI still
            // works for this session and the user can re-pick later.
        }
        DocumentFile doc = DocumentFile.fromTreeUri(getContext(), uri);
        String label = doc != null && doc.getName() != null ? doc.getName() : uri.getLastPathSegment();
        prefs().edit().putString(KEY_URI, uri.toString()).putString(KEY_LABEL, label).apply();
        JSObject ret = new JSObject();
        ret.put("uri", uri.toString());
        ret.put("label", label);
        call.resolve(ret);
    }

    @PluginMethod
    public void getFolder(PluginCall call) {
        JSObject ret = new JSObject();
        String uri = prefs().getString(KEY_URI, null);
        if (uri == null) {
            ret.put("uri", JSObject.NULL);
        } else {
            ret.put("uri", uri);
            ret.put("label", prefs().getString(KEY_LABEL, uri));
        }
        call.resolve(ret);
    }

    @PluginMethod
    public void readFile(PluginCall call) {
        String name = call.getString("name");
        JSObject ret = new JSObject();
        try {
            DocumentFile file = findFile(name);
            if (file == null) {
                ret.put("data", JSObject.NULL);
            } else {
                ret.put("data", Base64.encodeToString(readAll(file), Base64.NO_WRAP));
            }
            call.resolve(ret);
        } catch (Exception e) {
            call.reject("read failed: " + e.getMessage(), "IO", e);
        }
    }

    @PluginMethod
    public void writeFile(PluginCall call) {
        String name = call.getString("name");
        String data = call.getString("data");
        if (name == null || data == null) {
            call.reject("need name and data", "BAD_INPUT");
            return;
        }
        try {
            DocumentFile dir = tree();
            if (dir == null) {
                call.reject("no save folder chosen", "NO_FOLDER");
                return;
            }
            byte[] bytes = Base64.decode(data, Base64.DEFAULT);
            String tmpName = name + ".tmp";

            DocumentFile tmp = findFile(tmpName);
            if (tmp == null) tmp = dir.createFile("application/octet-stream", tmpName);
            if (tmp == null) throw new java.io.IOException("could not create " + tmpName);
            try (OutputStream out = getContext().getContentResolver()
                    .openOutputStream(tmp.getUri(), "wt")) {
                if (out == null) throw new java.io.IOException("no output stream for " + tmpName);
                out.write(bytes);
            }

            DocumentFile target = findFile(name);
            if (target != null) target.delete();
            if (!tmp.renameTo(name)) {
                // renameTo can fail on some providers; fall back to a move.
                DocumentFile fallback = dir.createFile("application/octet-stream", name);
                try (OutputStream out = getContext().getContentResolver()
                        .openOutputStream(fallback.getUri(), "wt")) {
                    out.write(bytes);
                }
                tmp.delete();
            }

            DocumentFile written = findFile(name);
            JSObject ret = new JSObject();
            ret.put("modified", written != null ? written.lastModified() : System.currentTimeMillis());
            call.resolve(ret);
        } catch (Exception e) {
            call.reject("write failed: " + e.getMessage(), "IO", e);
        }
    }

    // ── "Save as" export (ACTION_CREATE_DOCUMENT) ────────────────
    // Bytes are stashed in pendingExportBytes because the write can only
    // happen after the picker returns a document URI.
    private byte[] pendingExportBytes;

    @PluginMethod
    public void exportFile(PluginCall call) {
        String name = call.getString("name");
        String data = call.getString("data");
        String mime = call.getString("mimeType", "application/octet-stream");
        if (name == null || data == null) {
            call.reject("need name and data", "BAD_INPUT");
            return;
        }
        try {
            pendingExportBytes = Base64.decode(data, Base64.DEFAULT);
        } catch (Exception e) {
            call.reject("bad base64 data", "BAD_INPUT");
            return;
        }
        Intent intent = new Intent(Intent.ACTION_CREATE_DOCUMENT);
        intent.addCategory(Intent.CATEGORY_OPENABLE);
        intent.setType(mime);
        intent.putExtra(Intent.EXTRA_TITLE, name); // suggested file name
        startActivityForResult(call, intent, "exportFileResult");
    }

    @ActivityCallback
    private void exportFileResult(PluginCall call, ActivityResult result) {
        if (call == null) return;
        try {
            Intent data = result.getData();
            Uri uri = data != null ? data.getData() : null;
            if (uri == null) {
                call.reject("Export cancelled", "CANCELLED");
                return;
            }
            try (OutputStream out = getContext().getContentResolver()
                    .openOutputStream(uri, "wt")) { // "wt" truncates an existing pick
                if (out == null) throw new java.io.IOException("no output stream for picked document");
                out.write(pendingExportBytes);
            }
            JSObject ret = new JSObject();
            ret.put("uri", uri.toString());
            call.resolve(ret);
        } catch (Exception e) {
            call.reject("export write failed: " + e.getMessage(), "IO", e);
        } finally {
            pendingExportBytes = null;
        }
    }

    @PluginMethod
    public void statFile(PluginCall call) {
        String name = call.getString("name");
        JSObject ret = new JSObject();
        DocumentFile file = findFile(name);
        ret.put("exists", file != null);
        ret.put("modified", file != null ? file.lastModified() : 0);
        ret.put("size", file != null ? file.length() : 0);
        call.resolve(ret);
    }

    private DocumentFile findFile(String name) {
        if (name == null) return null;
        DocumentFile dir = tree();
        return dir != null ? dir.findFile(name) : null;
    }

    private byte[] readAll(DocumentFile file) throws java.io.IOException {
        try (InputStream in = getContext().getContentResolver().openInputStream(file.getUri())) {
            ByteArrayOutputStream buf = new ByteArrayOutputStream();
            byte[] chunk = new byte[64 * 1024];
            int n;
            while ((n = in.read(chunk)) != -1) buf.write(chunk, 0, n);
            return buf.toByteArray();
        }
    }
}

File: ios-overrides/SaveFolderPlugin.swift
// Capacitor plugin: let the user pick a folder (Files app document picker)
// and read/write individual files inside it. Memki keeps its whole
// collection in `memki.json` there so external tools can sync it.
//
// iOS port of android/.../SaveFolderPlugin.java — identical JS contract
// (see web/native-bridge.js), so the web side needs no platform branches:
//
//   pickFolder()            -> { uri, label }          rejects "CANCELLED"
//   getFolder()             -> { uri, label } | { uri: null }
//   readFile({ name })      -> { data: base64 | null }
//   writeFile({ name, data: base64 }) -> { modified }  (atomic tmp+rename)
//   statFile({ name })      -> { exists, modified, size }
//   readMedia({ name })     -> { data: base64 | null }   } media lives in a
//   writeMedia({ name, data: base64 }) -> { modified }   } "memki.media"
//                                                          } subfolder
//   exportFile({ name, data: base64, mimeType }) -> { uri }
//                             "save as" picker (UIDocumentPicker export);
//                             rejects "CANCELLED"
//
// Platform mapping:
//   - Android's persisted SAF tree URI becomes an iOS security-scoped
//     bookmark (Data) stored in UserDefaults. Resolving it yields a URL we
//     must startAccessingSecurityScopedResource() on before any I/O.
//   - `uri` in resolve payloads is the folder URL's absoluteString. It is
//     display/debug only — the bookmark data is the source of truth.
//   - `modified` is milliseconds since 1970, matching File.lastModified.
//
// Registration: Capacitor 6 auto-registers every CAPBridgedPlugin found in
// the app binary at bridge setup, so simply compiling this file into the
// App target (via ios-overrides/inject-into-xcodeproj.rb) is enough.

import Foundation
import Capacitor
import UIKit
import UniformTypeIdentifiers

@objc(SaveFolderPlugin)
public class SaveFolderPlugin: CAPPlugin, CAPBridgedPlugin, UIDocumentPickerDelegate {
    public let identifier = "SaveFolderPlugin"
    public let jsName = "SaveFolder"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "pickFolder", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getFolder", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "readFile", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "writeFile", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "statFile", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "readMedia", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "writeMedia", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "migrate", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "exportFile", returnType: CAPPluginReturnPromise),
    ]

    private static let bookmarkKey = "SaveFolder.bookmark"
    private static let labelKey = "SaveFolder.label"
    private static let mediaDirName = "memki.media"

    private let defaults = UserDefaults.standard

    // Only one system picker can be meaningfully in flight at a time (the
    // JS side serializes these calls, same as on Android).
    private enum PendingPicker {
        case pickFolder(CAPPluginCall)
        case exportFile(CAPPluginCall, tempURL: URL)
    }
    private var pendingPicker: PendingPicker?

    // ── Folder resolution (security-scoped bookmark) ──────────────

    /// Resolve the persisted bookmark to a folder URL, refreshing the
    /// stored bookmark when the system reports it stale. Returns nil when
    /// no folder was ever chosen or the bookmark no longer resolves (e.g.
    /// the folder was deleted) — the JS side treats that as "no folder"
    /// and shows the pick gate, mirroring a revoked SAF grant on Android.
    private func resolveFolder() -> URL? {
        guard let data = defaults.data(forKey: Self.bookmarkKey) else { return nil }
        do {
            var stale = false
            let url = try URL(resolvingBookmarkData: data, options: [], relativeTo: nil, bookmarkDataIsStale: &stale)
            if stale, let fresh = try? url.bookmarkData(options: [], includingResourceValuesForKeys: nil, relativeTo: nil) {
                defaults.set(fresh, forKey: Self.bookmarkKey)
            }
            return url
        } catch {
            return nil
        }
    }

    /// Run `body` with security scope held on the resolved folder.
    private func withFolder<T>(_ body: (URL) throws -> T) throws -> T {
        guard let url = resolveFolder() else {
            throw PluginError.noFolder
        }
        let accessing = url.startAccessingSecurityScopedResource()
        defer { if accessing { url.stopAccessingSecurityScopedResource() } }
        return try body(url)
    }

    private enum PluginError: Error, CustomStringConvertible {
        case noFolder
        var description: String { "no save folder chosen" }
    }

    // ── pickFolder / getFolder ────────────────────────────────────

    @objc func pickFolder(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            guard self.pendingPicker == nil, let presenter = self.bridge?.viewController else {
                call.reject("a picker is already active", "BUSY")
                return
            }
            let picker: UIDocumentPickerViewController
            if #available(iOS 14.0, *) {
                picker = UIDocumentPickerViewController(forOpeningContentTypes: [.folder], asCopy: false)
            } else {
                picker = UIDocumentPickerViewController(documentTypes: ["public.folder"], in: .open)
            }
            picker.allowsMultipleSelection = false
            picker.delegate = self
            self.pendingPicker = .pickFolder(call)
            presenter.present(picker, animated: true)
        }
    }

    @objc func getFolder(_ call: CAPPluginCall) {
        guard let url = resolveFolder() else {
            call.resolve(["uri": NSNull()])
            return
        }
        call.resolve([
            "uri": url.absoluteString,
            "label": defaults.string(forKey: Self.labelKey) ?? url.lastPathComponent,
        ])
    }

    // ── read / write / stat ───────────────────────────────────────

    @objc func readFile(_ call: CAPPluginCall) {
        guard let name = call.getString("name") else {
            call.reject("need name", "BAD_INPUT")
            return
        }
        do {
            let data = try withFolder { folder -> String? in
                let file = folder.appendingPathComponent(name)
                guard FileManager.default.fileExists(atPath: file.path) else { return nil }
                return try Data(contentsOf: file).base64EncodedString()
            }
            if let data = data {
                call.resolve(["data": data])
            } else {
                call.resolve(["data": NSNull()])
            }
        } catch PluginError.noFolder {
            call.resolve(["data": NSNull()]) // mirrors Android: findFile on null tree -> null
        } catch {
            call.reject("read failed: \(error.localizedDescription)", "IO")
        }
    }

    @objc func writeFile(_ call: CAPPluginCall) {
        guard let name = call.getString("name"), let data = call.getString("data") else {
            call.reject("need name and data", "BAD_INPUT")
            return
        }
        guard let bytes = Data(base64Encoded: data, options: .ignoreUnknownCharacters) else {
            call.reject("bad base64 data", "BAD_INPUT")
            return
        }
        do {
            let modified = try withFolder { folder -> Int64 in
                try Self.atomicWrite(bytes, to: folder.appendingPathComponent(name))
                return Self.modifiedMillis(of: folder.appendingPathComponent(name))
            }
            call.resolve(["modified": modified])
        } catch PluginError.noFolder {
            call.reject(PluginError.noFolder.description, "NO_FOLDER")
        } catch {
            call.reject("write failed: \(error.localizedDescription)", "IO")
        }
    }

    @objc func statFile(_ call: CAPPluginCall) {
        guard let name = call.getString("name") else {
            call.reject("need name", "BAD_INPUT")
            return
        }
        // stat is best-effort metadata: no folder / missing file resolves
        // to exists:false rather than rejecting (matches the Java side,
        // which never rejects from statFile).
        let result: [String: Any] = (try? withFolder { folder -> [String: Any] in
            let file = folder.appendingPathComponent(name)
            guard FileManager.default.fileExists(atPath: file.path) else {
                return ["exists": false, "modified": 0, "size": 0]
            }
            return [
                "exists": true,
                "modified": Self.modifiedMillis(of: file),
                "size": Self.fileSize(of: file),
            ]
        }) ?? ["exists": false, "modified": 0, "size": 0]
        call.resolve(result)
    }

    // ── Media files (memki.media/ subfolder of the save folder) ──

    @objc func readMedia(_ call: CAPPluginCall) {
        guard let name = call.getString("name") else {
            call.reject("need name", "BAD_INPUT")
            return
        }
        do {
            let data = try withFolder { folder -> String? in
                let file = folder.appendingPathComponent(Self.mediaDirName).appendingPathComponent(name)
                guard FileManager.default.fileExists(atPath: file.path) else { return nil }
                return try Data(contentsOf: file).base64EncodedString()
            }
            if let data = data {
                call.resolve(["data": data])
            } else {
                call.resolve(["data": NSNull()])
            }
        } catch PluginError.noFolder {
            call.resolve(["data": NSNull()])
        } catch {
            call.reject("media read failed: \(error.localizedDescription)", "IO")
        }
    }

    @objc func writeMedia(_ call: CAPPluginCall) {
        guard let name = call.getString("name"), let data = call.getString("data") else {
            call.reject("need name and data", "BAD_INPUT")
            return
        }
        guard let bytes = Data(base64Encoded: data, options: .ignoreUnknownCharacters) else {
            call.reject("bad base64 data", "BAD_INPUT")
            return
        }
        do {
            let modified = try withFolder { folder -> Int64 in
                let dir = folder.appendingPathComponent(Self.mediaDirName)
                try FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
                let target = dir.appendingPathComponent(name)
                try Self.atomicWrite(bytes, to: target)
                return Self.modifiedMillis(of: target)
            }
            call.resolve(["modified": modified])
        } catch PluginError.noFolder {
            call.reject(PluginError.noFolder.description, "NO_FOLDER")
        } catch {
            call.reject("media write failed: \(error.localizedDescription)", "IO")
        }
    }

    // ── One-time rename from the pre-rename (oss-anki.*) names ────

    @objc func migrate(_ call: CAPPluginCall) {
        try? withFolder { folder in
            renameIfOrphan(in: folder, from: "oss-anki.json", to: "memki.json")
            renameIfOrphan(in: folder, from: "oss-anki-backup.json", to: "memki-backup.json")
            renameIfOrphan(in: folder, from: "oss-anki.media", to: Self.mediaDirName)
        }
        call.resolve()
    }

    private func renameIfOrphan(in folder: URL, from oldName: String, to newName: String) {
        let fm = FileManager.default
        let oldURL = folder.appendingPathComponent(oldName)
        let newURL = folder.appendingPathComponent(newName)
        if fm.fileExists(atPath: oldURL.path) && !fm.fileExists(atPath: newURL.path) {
            try? fm.moveItem(at: oldURL, to: newURL)
        }
    }

    // ── "Save as" export (UIDocumentPicker export mode) ───────────

    @objc func exportFile(_ call: CAPPluginCall) {
        guard let name = call.getString("name"), let data = call.getString("data") else {
            call.reject("need name and data", "BAD_INPUT")
            return
        }
        guard let bytes = Data(base64Encoded: data, options: .ignoreUnknownCharacters) else {
            call.reject("bad base64 data", "BAD_INPUT")
            return
        }
        do {
            // The export picker needs an on-disk source URL it can copy
            // out of the sandbox; stage the bytes in a unique temp dir.
            let staging = FileManager.default.temporaryDirectory
                .appendingPathComponent("memki-export-\(UUID().uuidString)")
            try FileManager.default.createDirectory(at: staging, withIntermediateDirectories: true)
            let tempURL = staging.appendingPathComponent(name)
            try bytes.write(to: tempURL)

            DispatchQueue.main.async {
                guard self.pendingPicker == nil, let presenter = self.bridge?.viewController else {
                    try? FileManager.default.removeItem(at: staging)
                    call.reject("a picker is already active", "BUSY")
                    return
                }
                let picker: UIDocumentPickerViewController
                if #available(iOS 14.0, *) {
                    picker = UIDocumentPickerViewController(forExporting: [tempURL], asCopy: true)
                } else {
                    picker = UIDocumentPickerViewController(urls: [tempURL], in: .exportToService)
                }
                picker.delegate = self
                self.pendingPicker = .exportFile(call, tempURL: tempURL)
                presenter.present(picker, animated: true)
            }
        } catch {
            call.reject("export staging failed: \(error.localizedDescription)", "IO")
        }
    }

    // ── UIDocumentPickerDelegate ──────────────────────────────────

    public func documentPicker(_ controller: UIDocumentPickerViewController, didPickDocumentsAt urls: [URL]) {
        guard let pending = pendingPicker else { return }
        pendingPicker = nil
        switch pending {
        case .pickFolder(let call):
            guard let url = urls.first else {
                call.reject("Folder pick cancelled", "CANCELLED")
                return
            }
            do {
                let bookmark = try url.bookmarkData(options: [], includingResourceValuesForKeys: nil, relativeTo: nil)
                defaults.set(bookmark, forKey: Self.bookmarkKey)
                defaults.set(url.lastPathComponent, forKey: Self.labelKey)
                call.resolve(["uri": url.absoluteString, "label": url.lastPathComponent])
            } catch {
                call.reject("could not persist folder access: \(error.localizedDescription)", "IO")
            }
        case .exportFile(let call, let tempURL):
            defer { try? FileManager.default.removeItem(at: tempURL.deletingLastPathComponent()) }
            guard let dest = urls.first else {
                call.reject("Export cancelled", "CANCELLED")
                return
            }
            call.resolve(["uri": dest.absoluteString])
        }
    }

    public func documentPickerWasCancelled(_ controller: UIDocumentPickerViewController) {
        guard let pending = pendingPicker else { return }
        pendingPicker = nil
        switch pending {
        case .pickFolder(let call):
            call.reject("Folder pick cancelled", "CANCELLED")
        case .exportFile(let call, let tempURL):
            try? FileManager.default.removeItem(at: tempURL.deletingLastPathComponent())
            call.reject("Export cancelled", "CANCELLED")
        }
    }

    // ── Helpers ───────────────────────────────────────────────────

    /// Write via a `<name>.tmp` sibling renamed over the target — a crash
    /// mid-write can never leave a truncated save file, and file watchers
    /// only ever see complete files. Same contract as the Android side.
    private static func atomicWrite(_ bytes: Data, to target: URL) throws {
        let fm = FileManager.default
        let tmp = target.deletingLastPathComponent()
            .appendingPathComponent(target.lastPathComponent + ".tmp")
        try bytes.write(to: tmp, options: [])
        if fm.fileExists(atPath: target.path) {
            try fm.removeItem(at: target)
        }
        try fm.moveItem(at: tmp, to: target)
    }

    private static func modifiedMillis(of url: URL) -> Int64 {
        let values = try? url.resourceValues(forKeys: [.contentModificationDateKey])
        return Int64((values?.contentModificationDate?.timeIntervalSince1970 ?? 0) * 1000)
    }

    private static func fileSize(of url: URL) -> Int64 {
        let values = try? url.resourceValues(forKeys: [.fileSizeKey])
        return Int64(values?.fileSize ?? 0)
    }
}
